packages/cron/src/jobs/measureNftTimeToRetrievability.js

```diff
@@ -0,0 +1,380 @@
+    metric.observe(retrieval.duration.toNumber() / 1000)
+    const pushAddArgs = {
+      jobName: metricsPushGatewayJobName,
+    }
+    await pushgateway.pushAdd(pushAddArgs)
+    options.log('debug', { type: 'pushgateway.pushAdd', args: pushAddArgs })
+  }
+  return push
+}
+
+/**
+ * Push retrieval metrics to a pushgateway
+ * @param {object} config
+ * @param {string} config.metricsPushGateway
+ * @param {HttpAuthorization} config.metricsPushGatewayAuthorization
+ * @param {Fetcher['fetch']} config.fetch
+ * @returns {RetrievalMetricsLogger}
+ */
+export function createFetchMetricsLogger(config) {
+  return async function (options, retrieval) {
+    /** @type [string, RequestInit] */
+    const pushRequest = [
+      config.metricsPushGateway,
+      {
+        method: 'post',
+        headers: {
+          authorization: config.metricsPushGatewayAuthorization.authorization,
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          'content-type': 'application/octet-stream',
+        },
+        body: formatRetrievalMetrics(retrieval),
+      },
+    ]
+    const pushResponse = await config.fetch.apply(null, pushRequest)
+    if (!pushResponse.ok) {
+      options.log('warn', {
+        type: 'error',
+        message: 'unsuccessful metrics push',
+        text: await pushResponse.text(),
+      })
+    }
+    assert.equal(
+      pushResponse.ok,
+      true,
+      'metrics push response code should indicate success'
+    )
+    options.log('info', {
+      type: 'metricsPushed',
+      target: config.metricsPushGateway,
+    })
+  }
+}
+
+/**
+ * @typedef BasicAuthOptions
+ * @property {string} basicAuth - base64-encoded user:pass
+ */
+
+/**
+ * @typedef UserPassOptions
+ * @property {string} username - username
+ * @property {string} password - password
+ */
+
+/**
+ * @param {BasicAuthOptions|UserPassOptions} options
+ * @returns {`Basic ${string}`} - http Authorization header value
+ */
+export function basicAuthorizationHeaderValue(options) {
+  const basicAuthValue =
+    'basicAuth' in options
+      ? options.basicAuth
+      : btoa(`${options.username}:${options.password}`)
+  /** @type {`Basic ${string}`} */
+  const headerValue = `Basic ${basicAuthValue}`
+  return headerValue
+}
+
+/**
+ * Format metrics about a retrieval into a format that a prometheus pushgateway expects.
+ * Example output: https://github.com/web3-storage/web3.storage/blob/main/.github/workflows/cron-test.yml#L38
+ * @param {RetrieveLog} retrieval
+ * @returns {string} prometheus metrics string
+ */
+function formatRetrievalMetrics(retrieval) {
+  const durationSeconds = retrieval.duration.size / 1000
+  return (
+    [
+      '# TYPE nftstorage_retrieval_duration_seconds gauge',
+      '# HELP nftstorage_retrieval_duration_seconds How long it took to retrieve an nft image',
+      `nftstorage_retrieval_duration_seconds{byteLength="${retrieval.contentLength}"} ${durationSeconds}`,
```

Comment:
@gobengo You need to move to a histogram. A gauge is the wrong measurement type for this metric. I outlined my recommendation in the slack thread on this PR but for convenience, tldr, remove the url and image field, `size` should be a set of buckets, i.e. 10MB, 25MB, 50MB, 100MB and the metric should be a histogram (https://www.npmjs.com/package/prom-client#Histogram)

Reply:
@lanzafame Thanks for the review. Helped me realize this code is now unused, as I switched to using npm `prom-client` per @alanshaw suggestion, and in doing so I switched to a [Histogram](https://github.com/nftstorage/nft.storage/pull/1945/files#diff-5823c1dd943cae9a4760e898922c5e7a0344a231d51560db2541882f2a9beb41R15).

> size should be a set of buckets, i.e. 10MB, 25MB, 50MB, 100MB

By size do you mean `byteLength`? fwiw I removed all mention of the fields. I think we're mostly interested in measuring the retrieval duration in seconds. The byteLength/size was a nice to have.

Thanks for educating me about the perf et al tradeoffs in storing extra fields and not using buckets.

Reply:
If the bytelength is still variable (as in the size of the file uploaded does change) then it should be recorded against the duration value so that the histogram can be presented correctly on the grafana end. Otherwise we will have uploads that take longer due to the file size but appear to be because the service is being slow.